RoddyCore/src/de/dkfz/roddy/execution/io/SSHExecutionService.groovy

```diff
@@ -353,83 +361,85 @@ class SSHExecutionService extends RemoteExecutionService {
-        SSHPoolConnectionSet connectionSet = waitForService()
-        SSHClient sshClient = connectionSet.client
+    private static final ExecutorService executorService = Executors.newCachedThreadPool()
 
-        if (waitFor) {
+    @Override
+    ExecutionResult _execute(String command,
+                             boolean waitFor = true,
+                             Duration timeout = Duration.ZERO,
+                             OutputStream outputStream = null) {
+        _execute(command, waitFor, timeout, outputStream, executorService)
+    }
+
+    /**
+     * An com.hierynomus:sshj:0.23.0 and JSCH Agent 0.0.9 (-> net.schmizz:sshj:0.10.0) compatibility
+     * problem prevents the usage of Command.join(long, TimeUnit) from working. The best solution is
+     * probably to get rid of the whole SSHJ/JSCH stack and instead use mina-sshd, which is not
+     * completely outdated and actively maintained (and can use the SSH-agent and execute with timeouts.
+     *
+     * The solution taken now uses the Linux command `timeout`.
+     *
+     * @param command
+     * @param waitFor
+     * @param timeout
+     * @param outputStream
+     * @param executorService
+     * @return
+     */
+    ExecutionResult _execute(String command,
+                             boolean waitFor = true,
+                             Duration timeout = Duration.ZERO,
+                             OutputStream outputStream = null,
+                             ExecutorService executorService) {
+
+        CompletableFuture<Tuple3<Integer,List<String>, List<String>>> processF = CompletableFuture.supplyAsync({
+            SSHPoolConnectionSet connectionSet = waitForService()
+            SSHClient sshClient = connectionSet.client
+
+            // Append a newgrp (Philip: better "newgrp -"!) to each command, so that all
+            // command context in the proper group context.
             connectionSet.acquire()
             Session session = sshClient.startSession()
-            Session.Command cmd
-
+            List<String> stdout
+            List<String> stderr
+            Integer exitCode
             try {
-                cmd = session.exec(command)
+                Session.Command executingCommand
+                if (timeout != Duration.ZERO) {
+                    // TODO Remove this workaround. Try switching to mina-sshd.
+                    Long seconds = Math.round(Math.ceil(timeout.toNanos() / 1000000000.0d))
+                    Long killSeconds = 60
+                    String timeoutCommand = "timeout --preserve-status -s TERM -k ${killSeconds}s ${seconds}s"
+                    executingCommand = session.exec("${timeoutCommand} ${command}")
+                } else {
+                    executingCommand = session.exec(command)
+                }
+                if (outputStream) {
+                    stdout = LocalExecutionHelper.readStringStream(executingCommand.inputStream, outputStream)
```

Comment:
Note that this implementation differs from the one in RoddyToolLib. The main reason is that I didn't know the details about output streaming with SSHJ. I therefore decided to basically stick to old (working) implementation with synchronous readout of output streams.